File: 174-dungeon-game/dungeon-game.cpp
class Solution {
public:

    vector<vector<int>>dp;

    int calculateMinimumHP(vector<vector<int>>& dungeon) {
        int n = dungeon.size();
        int m = dungeon[0].size();
        dp.assign(n, vector<int>(m,0));
        dp[n-1][m-1] = max(1, 1-dungeon[n-1][m-1]);
        for(int i = n-1; i>=0; i--){
            for(int j = m-1; j>=0; j--){
                if(i == n-1 && j == m-1) continue;
                int right,down;
                if(j+1<m) right = dp[i][j+1];
                else right = INT_MAX;
                if(i+1<n) down = dp[i+1][j];
                else down = INT_MAX;
                int need = min(right,down) - dungeon[i][j];
                dp[i][j] = max(1, need);
            }
        }
        return dp[0][0];
    }
};

File: 174-dungeon-game/dungeon-game.js
/**
 * @param {number[][]} dungeon
 * @return {number}
 */
var calculateMinimumHP = function(dungeon) {
    const n = dungeon.length;
    const m = dungeon[0].length;
    const dp = Array.from({length: n}, () => Array(m).fill(-1));
    function solve(i,j){
        if(i>=n || j>=m) return Infinity;
        if(i === n-1 && j === m-1){
            return Math.max(1,1-dungeon[i][j]);
        }
        if(dp[i][j]!== -1) return dp[i][j];

        const down = solve(i+1, j);
        const right = solve(i, j+1);
        const need = Math.min(down, right);
        dp[i][j] = Math.max(1, need - dungeon[i][j]);
        return dp[i][j];
    }
    return solve(0,0);
};